File: app/src/main/java/com/blankj/androidutilcode/feature/core/snackbar/SnackbarActivity.java
package com.blankj.androidutilcode.feature.core.snackbar;

import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.StringRes;
import android.text.SpannableStringBuilder;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.blankj.androidutilcode.R;
import com.blankj.androidutilcode.base.BaseBackActivity;
import com.blankj.utilcode.util.SnackbarUtils;
import com.blankj.utilcode.util.SpanUtils;
import com.blankj.utilcode.util.ToastUtils;

/**
 * <pre>
 *     author: Blankj
 *     blog  : http://blankj.com
 *     time  : 2016/10/17
 *     desc  : Snackbar 工具类 Demo
 * </pre>
 */
public class SnackbarActivity extends BaseBackActivity {

    View snackBarRootView;

    public static void start(Context context) {
        Intent starter = new Intent(context, SnackbarActivity.class);
        context.startActivity(starter);
    }

    @Override
    public void initData(Bundle bundle) {

    }

    @Override
    public int bindLayout() {
        return R.layout.activity_snackbar;
    }

    @Override
    public void initView(Bundle savedInstanceState, View view) {
        getToolBar().setTitle(getString(R.string.demo_snackbar));

        snackBarRootView = findViewById(android.R.id.content);
        findViewById(R.id.btn_short_snackbar).setOnClickListener(this);
        findViewById(R.id.btn_short_snackbar_with_action).setOnClickListener(this);
        findViewById(R.id.btn_long_snackbar).setOnClickListener(this);
        findViewById(R.id.btn_long_snackbar_with_action).setOnClickListener(this);
        findViewById(R.id.btn_indefinite_snackbar).setOnClickListener(this);
        findViewById(R.id.btn_indefinite_snackbar_with_action).setOnClickListener(this);
        findViewById(R.id.btn_add_view).setOnClickListener(this);
        findViewById(R.id.btn_add_view_with_action).setOnClickListener(this);
        findViewById(R.id.btn_show_success).setOnClickListener(this);
        findViewById(R.id.btn_show_warning).setOnClickListener(this);
        findViewById(R.id.btn_show_error).setOnClickListener(this);
        findViewById(R.id.btn_dismiss_snackbar).setOnClickListener(this);
    }

    @Override
    public void doBusiness() {

    }

    @Override
    public void onWidgetClick(View view) {
        switch (view.getId()) {
            case R.id.btn_short_snackbar:
                SnackbarUtils.with(snackBarRootView)
                        .setMessage(getMsg(R.string.snackbar_short))
                        .setMessageColor(Color.WHITE)
                        .setBgResource(R.drawable.shape_top_round_rect)
                        .show();
                break;

            case R.id.btn_short_snackbar_with_action:
                SnackbarUtils.with(snackBarRootView)
                        .setMessage(getMsg(R.string.snackbar_short))
                        .setMessageColor(Color.WHITE)
                        .setBgResource(R.drawable.shape_top_round_rect)
                        .setAction(getString(R.string.snackbar_click), Color.YELLOW, new View.OnClickListener() {
                            @Override
                            public void onClick(View v) {
                                ToastUtils.showShort(getString(R.string.snackbar_click));
                            }
                        })
                        .show();
                break;

            case R.id.btn_long_snackbar:
                SnackbarUtils.with(snackBarRootView)
                        .setMessage(getMsg(R.string.snackbar_long))
                        .setMessageColor(Color.WHITE)
                        .setDuration(SnackbarUtils.LENGTH_LONG)
                        .setBgResource(R.drawable.shape_top_round_rect)
                        .show();
                break;

            case R.id.btn_long_snackbar_with_action:
                SnackbarUtils.with(snackBarRootView)
                        .setMessage(getMsg(R.string.snackbar_long))
                        .setMessageColor(Color.WHITE)
                        .setBgResource(R.drawable.shape_top_round_rect)
                        .setDuration(SnackbarUtils.LENGTH_LONG)
                        .setAction(getString(R.string.snackbar_click), Color.YELLOW, new View.OnClickListener() {
                            @Override
                            public void onClick(View v) {
                                ToastUtils.showShort(getString(R.string.snackbar_click));
                            }
                        })
                        .show();
                break;

            case R.id.btn_indefinite_snackbar:
                SnackbarUtils.with(snackBarRootView)
                        .setMessage(getMsg(R.string.snackbar_indefinite))
                        .setMessageColor(Color.WHITE)
                        .setDuration(SnackbarUtils.LENGTH_INDEFINITE)
                        .setBgResource(R.drawable.shape_top_round_rect)
                        .show();
                break;

            case R.id.btn_indefinite_snackbar_with_action:
                SnackbarUtils.with(snackBarRootView)
                        .setMessage(getMsg(R.string.snackbar_indefinite))
                        .setMessageColor(Color.WHITE)
                        .setDuration(SnackbarUtils.LENGTH_INDEFINITE)
                        .setBgResource(R.drawable.shape_top_round_rect)
                        .setAction(getString(R.string.snackbar_click), Color.YELLOW, new View.OnClickListener() {
                            @Override
                            public void onClick(View v) {
                                ToastUtils.showShort(getString(R.string.snackbar_click));
                            }
                        })
                        .show();
                break;

            case R.id.btn_add_view:
                ViewGroup.LayoutParams params = new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);
                SnackbarUtils.with(snackBarRootView)
                        .setBgColor(Color.TRANSPARENT)
                        .setDuration(SnackbarUtils.LENGTH_INDEFINITE)
                        .show();
                SnackbarUtils.addView(R.layout.snackbar_custom, params);
                break;

            case R.id.btn_add_view_with_action:
                SnackbarUtils.with(snackBarRootView)
                        .setBgColor(Color.TRANSPARENT)
                        .setDuration(SnackbarUtils.LENGTH_INDEFINITE)
                        .show();
                params = new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                        ViewGroup.LayoutParams.MATCH_PARENT);
                SnackbarUtils.addView(R.layout.snackbar_custom, params);
                View snackbarView = SnackbarUtils.getView();
                if (snackbarView != null) {
                    TextView tvSnackbarCustom = snackbarView.findViewById(R.id.tv_snackbar_custom);
                    tvSnackbarCustom.setText("点我可消失");
                    snackbarView.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            SnackbarUtils.dismiss();
                        }
                    });
                }
                break;

            case R.id.btn_show_success:
                SnackbarUtils.with(snackBarRootView)
                        .setMessage(getMsg(R.string.snackbar_success))
                        .showSuccess();
                break;

            case R.id.btn_show_warning:
                SnackbarUtils.with(snackBarRootView)
                        .setMessage(getMsg(R.string.snackbar_warning))
                        .showWarning();
                break;

            case R.id.btn_show_error:
                SnackbarUtils.with(snackBarRootView)
                        .setMessage(getMsg(R.string.snackbar_error))
                        .showError();
                break;

            case R.id.btn_dismiss_snackbar:
                SnackbarUtils.dismiss();
                break;
        }
    }

    private SpannableStringBuilder getMsg(@StringRes int resId) {
        return new SpanUtils()
                .appendImage(R.mipmap.ic_launcher, SpanUtils.ALIGN_CENTER)
                .appendSpace(32)
                .append(getString(resId)).setFontSize(24, true)
                .create();
    }
}

File: feature/utilcode/pkg/src/main/java/com/blankj/utilcode/pkg/feature/snackbar/SnackbarActivity.kt
package com.blankj.utilcode.pkg.feature.snackbar

import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.support.annotation.StringRes
import android.text.SpannableStringBuilder
import android.view.ViewGroup
import android.widget.TextView
import com.blankj.common.activity.CommonActivity
import com.blankj.common.item.CommonItem
import com.blankj.common.item.CommonItemClick
import com.blankj.utilcode.pkg.R
import com.blankj.utilcode.util.CollectionUtils
import com.blankj.utilcode.util.SnackbarUtils
import com.blankj.utilcode.util.SpanUtils
import com.blankj.utilcode.util.ToastUtils

/**
 * ```
 * author: Blankj
 * blog  : http://blankj.com
 * time  : 2016/10/17
 * desc  : demo about SnackbarUtils
 * ```
 */
class SnackbarActivity : CommonActivity() {

    companion object {
        fun start(context: Context) {
            val starter = Intent(context, SnackbarActivity::class.java)
            context.startActivity(starter)
        }
    }

    override fun bindTitleRes(): Int {
        return R.string.demo_snackbar
    }

    override fun bindItems(): MutableList<CommonItem<*>> {
        return CollectionUtils.newArrayList(
                CommonItemClick(R.string.snackbar_show_short) {
                    SnackbarUtils.with(mContentView)
                            .setMessage(getMsg(R.string.snackbar_short))
                            .setMessageColor(Color.WHITE)
                            .setBgResource(R.drawable.snackbar_custom_bg)
                            .show()
                },
                CommonItemClick(R.string.snackbar_show_short_with_action) {
                    SnackbarUtils.with(mContentView)
                            .setMessage(getMsg(R.string.snackbar_short))
                            .setMessageColor(Color.WHITE)
                            .setBgResource(R.drawable.snackbar_custom_bg)
                            .setAction(getString(R.string.snackbar_click), Color.YELLOW) { ToastUtils.showShort(getString(R.string.snackbar_click)) }
                            .show()
                },
                CommonItemClick(R.string.snackbar_show_long) {
                    SnackbarUtils.with(mContentView)
                            .setMessage(getMsg(R.string.snackbar_long))
                            .setMessageColor(Color.WHITE)
                            .setDuration(SnackbarUtils.LENGTH_LONG)
                            .setBgResource(R.drawable.snackbar_custom_bg)
                            .show()
                },
                CommonItemClick(R.string.snackbar_show_long_with_action) {
                    SnackbarUtils.with(mContentView)
                            .setMessage(getMsg(R.string.snackbar_long))
                            .setMessageColor(Color.WHITE)
                            .setBgResource(R.drawable.snackbar_custom_bg)
                            .setDuration(SnackbarUtils.LENGTH_LONG)
                            .setAction(getString(R.string.snackbar_click), Color.YELLOW) { ToastUtils.showShort(getString(R.string.snackbar_click)) }
                            .show()
                },
                CommonItemClick(R.string.snackbar_show_indefinite) {
                    SnackbarUtils.with(mContentView)
                            .setMessage(getMsg(R.string.snackbar_indefinite))
                            .setMessageColor(Color.WHITE)
                            .setDuration(SnackbarUtils.LENGTH_INDEFINITE)
                            .setBgResource(R.drawable.snackbar_custom_bg)
                            .show()
                },
                CommonItemClick(R.string.snackbar_show_indefinite_with_action) {
                    SnackbarUtils.with(mContentView)
                            .setMessage(getMsg(R.string.snackbar_indefinite))
                            .setMessageColor(Color.WHITE)
                            .setDuration(SnackbarUtils.LENGTH_INDEFINITE)
                            .setBgResource(R.drawable.snackbar_custom_bg)
                            .setAction(getString(R.string.snackbar_click), Color.YELLOW) { ToastUtils.showShort(getString(R.string.snackbar_click)) }
                            .show()
                },
                CommonItemClick(R.string.snackbar_add_view) {
                    val params = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
                    SnackbarUtils.with(mContentView)
                            .setBgColor(Color.TRANSPARENT)
                            .setDuration(SnackbarUtils.LENGTH_INDEFINITE)
                            .show()
                    SnackbarUtils.addView(R.layout.snackbar_custom, params)
                },
                CommonItemClick(R.string.snackbar_add_view_with_action) {
                    val params: ViewGroup.LayoutParams = ViewGroup.LayoutParams(
                            ViewGroup.LayoutParams.MATCH_PARENT,
                            ViewGroup.LayoutParams.MATCH_PARENT
                    )
                    SnackbarUtils.with(mContentView)
                            .setBgColor(Color.TRANSPARENT)
                            .setDuration(SnackbarUtils.LENGTH_INDEFINITE)
                            .show()
                    SnackbarUtils.addView(R.layout.snackbar_custom, params)
                    val snackbarView = SnackbarUtils.getView()
                    if (snackbarView != null) {
                        val tvSnackbarCustom = snackbarView.findViewById<TextView>(R.id.snackbarCustomTv)
                        tvSnackbarCustom.setText(R.string.snackbar_click_to_dismiss)
                        snackbarView.setOnClickListener { SnackbarUtils.dismiss() }
                    }
                },
                CommonItemClick(R.string.snackbar_show_success) {
                    SnackbarUtils.with(mContentView)
                            .setMessage(getMsg(R.string.snackbar_success))
                            .showSuccess()
                },
                CommonItemClick(R.string.snackbar_show_warning) {
                    SnackbarUtils.with(mContentView)
                            .setMessage(getMsg(R.string.snackbar_warning))
                            .showWarning()
                },
                CommonItemClick(R.string.snackbar_show_error) {
                    SnackbarUtils.with(mContentView)
                            .setMessage(getMsg(R.string.snackbar_error))
                            .showError()
                },
                CommonItemClick(R.string.snackbar_dismiss) {
                    SnackbarUtils.dismiss()
                }
        )
    }

    private fun getMsg(@StringRes resId: Int): SpannableStringBuilder {
        return SpanUtils()
                .appendImage(R.mipmap.ic_launcher, SpanUtils.ALIGN_CENTER)
                .appendSpace(32)
                .append(getString(resId)).setFontSize(24, true)
                .create()
    }
}
